python/remotivelabs-broker/remotivelabs/broker/sync/client.py

```diff
@@ -0,0 +1,263 @@
+                url: str,
+                api_key: Union[str, None] = None, ):
+        self.url = url
+        self.api_key = api_key
+        if url.startswith("https"):
+            if api_key is None:
+                raise BrokerException("You must supply api-key or access-token to use a cloud broker")
+            self._intercept_channel = br.create_channel(url, self.api_key, None)
+        else:
+            self._intercept_channel = br.create_channel(url, None, None)
+
+        self._network_stub = network_api_pb2_grpc.NetworkServiceStub(self._intercept_channel)
+        self._system_stub = system_api_pb2_grpc.SystemServiceStub(self._intercept_channel)
+        self._traffic_stub = traffic_api_pb2_grpc.TrafficServiceStub(self._intercept_channel)
+        self._signal_creator = SignalCreator(self._system_stub)
+        if self.on_connect is not None:
+            self.on_connect(self)
+
+    def _validate_and_get_subscribed_signals(self, subscribed_namespaces: List[str], subscribed_signals: List[str]) \
+            -> List[SignalIdentifier]:
+        # Since we cannot know which list[signals] belongs to which namespace we need to fetch
+        # all signals from the broker and find the proper signal with namespace. Finally,  we
+        # also filter out namespaces that we do not need since we might have duplicated signal names
+        # over namespaces
+        # Begin
+
+        def verify_namespace(available_signal: SignalIdentifier):
+            return list(filter(lambda namespace: available_signal.namespace == namespace, subscribed_namespaces))
+
+        def find_subscribed_signal(available_signal: SignalIdentifier):
+            return list(filter(lambda s: available_signal.name == s, subscribed_signals))
+
+        available_signals: list[SignalIdentifier] = list(filter(verify_namespace, self.list_signal_names()))
+        signals_to_subscribe_to: list[SignalIdentifier] = list(filter(find_subscribed_signal, available_signals))
+
+        # Check if subscription is done on signal that is not in any of these namespaces
+        signals_subscribed_to_but_does_not_exist = \
+            set(subscribed_signals) - set(map(lambda s: s.name, signals_to_subscribe_to))
+
+        if len(signals_subscribed_to_but_does_not_exist) > 0:
+            raise BrokerException(f"One or more signals you subscribed to does not exist "
+                                  f", {signals_subscribed_to_but_does_not_exist}")
+
+        return list(map(lambda s: SignalIdentifier(s.name, s.namespace), signals_to_subscribe_to))
+
+    def subscribe(self,
+                  signal_names: list[str],
+                  namespaces: list[str],
+                  on_signals: Callable[[SignalsInFrame], None] = None,
+                  changed_values_only: bool = True):
+
+        if on_signals is None and self.on_signals is None:
+            print("You have not specified global client.on_signals nor client.subscribe(on_signals=callback)", file=sys.stderr)
```

Comment:
Should we allow this or should we return?

Reply:
raise error is probably better!